File: src/gp/gpmempool.cpp
/*
 * Copyright (C) 2010 Telmo Menezes.
 * telmo@telmomenezes.com
 */


#include "gpmempool.h"


GPMemPool::GPMemPool(size_t buffer_size)
{
    _first_gpnode = NULL;
    _buffer_size = buffer_size;
    add_buffer();
}


GPMemPool::~GPMemPool()
{
    for (list<GPNode*>::iterator iterBuff = _buffers.begin();
            iterBuff != _buffers.end();
            iterBuff++)
    {
        free(*iterBuff);
    }
}


GPNode* GPMemPool::get_node()
{
    GPNode* node;
    if (_first_gpnode) {
        node = _first_gpnode;
        _first_gpnode = node->parent;
        return node;
    }
    else {
        add_buffer();
        return get_node();
    }
}


void GPMemPool::return_node(GPNode* node)
{
    node->parent = _first_gpnode;
    _first_gpnode = node;
}


void GPMemPool::add_buffer()
{
    GPNode* buffer = (GPNode *)malloc(sizeof(GPNode) * _buffer_size);
    for (unsigned int i = 0; i < _buffer_size; i++)
        if (i == 0)
            buffer[i].parent = NULL;
        else
            buffer[i].parent = &(buffer[i - 1]);

    _first_gpnode = &(buffer[_buffer_size - 1]);
    _buffers.push_back(buffer);
}

File: src/com/telmomenezes/synthetic/gp/GPMemPool.java
package com.telmomenezes.synthetic.gp;


import java.util.ArrayList;


/**
 * Memory pool of genetic program nodes.
 * 
 * Allows for the reuse of GPNode objects. Since these objects are created
 * and destroyed in large quantities during the evolutionary process, creating
 * a pool for their reuse is a relevant optimization.
 * 
 * @author Telmo Menezes (telmo@telmomenezes.com)
 */
public class GPMemPool {


	private static GPMemPool _instance = null;

    private ArrayList<GPNode> buffer;
    

	private GPMemPool(int initial_buffer_size)
	{
		buffer = new ArrayList<GPNode>(initial_buffer_size);
	}


	static GPMemPool instance()
	{
		if (_instance == null)
			_instance = new GPMemPool(100000);

		return _instance;
	}


	public GPNode getNode()
	{
		GPNode node;
		
		if (buffer.size() == 0) {
			node = new GPNode();
		}
		else {
			node = buffer.get(0);
			buffer.remove(0);
			
		}
		
		return node;
	}


	public void returnNode(GPNode node)
	{
		buffer.add(node);
	}
}